HPCC-30365 Add XREF Sasha service to K8s

jackdelv opened the pull request:
These changes add Xref functionality to the containerized version of the platform. There were a couple changes because thor file parts were striped across multiple directories.

I'm not sure that I correctly added the new XRef pod because everything still runs on the ECL Watch pod.

What is the best way to make this toggleable? An additional #ifdef? Should I still be using _CONTAINERIZED?

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [ ] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [x] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

jackdelv committed:
HPCC-30365 Add XREF Sasha service to K8s

jakesmith commented on dali/dfuXRefLib/dfuxreflib.cpp:
```diff
@@ -1867,7 +1879,11 @@ class CXRefManager: public CXRefManagerBase
+                        unsigned short port = 0; // No need to connect to Dafs, data is available through mounted storage planes
+#else
                         unsigned short port = getDafsPort(node.endpoint(),numfails,&crit);
+#endif
```
I think this should instead check to see if node is localhost, before setting to 0.
i.e. in BM too, if BM, we don't want to go via dafilesrv really. In cloud we *do* want to go via dafilesrv if the file is associated with a non-local hosts. That is not typical, but planes can be configured with hosted storage, so we need to continue to support access via dafielsrv if needed.

jackdelv replied:
Fixed to check if the endpoint is localhost.

jakesmith commented on dali/dfuXRefLib/dfuxreflib.cpp:
```diff
@@ -2600,7 +2616,11 @@ class CXRefManager: public CXRefManagerBase
         Owned<IGroup> g;
         unsigned j;
         if (!nclusters) {
+#ifdef _CONTAINERIZED
```
preferable to use isCotainerized()

In fact, I think it would be worth just changing the message to the planes version.
In BM, internally, we generate planes for all clusters, such that methods like getStoragePlane/getPlanesIterator work with clusters.

jackdelv replied:
Changed to planes message.

jakesmith commented on dali/dfuXRefLib/dfuxreflib.cpp:
```diff
@@ -2618,18 +2638,29 @@ class CXRefManager: public CXRefManagerBase
+
+            loadFromDFS(*this,g,1,storageDir,storagePlane);
+            xrefRemoteDirectories(g,numdirs,storageDir,numthreads);
+        }
```
I think this ok, processing 1 {plane + dir} at a time, vs BM, which is processing all at once.
However, 'g' is a melded group of all, so **if** there were non-local storage planes (hostGroup storage) then 'g' would be a combo of all, but each call here is dealing with 1.
So the 'g' combine code above should be moved to BM only, and this should use queryNamedGroupStore().lookup(clusters[i]); directly

jakesmith commented on dali/dfuXRefLib/dfuxreflib.cpp:
```diff
@@ -2669,24 +2700,31 @@ IPropertyTree *  runXRef(unsigned nclusters,const char **clusters,IXRefProgressC
+    for (int i = 0; i < numdirs; i++)
+    {
+        Owned<IPropertyTree> storagePlane = getStoragePlane(clusters[i]);
+        dirs[i] = storagePlane->queryProp("@prefix");
```
I am not sure it makes sense to be enumerating the directories into 'dirs' and passing them xrefmanager.process.

In BM, the dirs are not different cluster dirs, but the primary/replicate dirs, which are consistent across all clusters.

I think we should keep it simple in cloud version, and assert that we are only dealing with 1 plane at a time (at least initially), the dirs should contain a single directory of the plane prefix being processes

jackdelv replied:
Changed to only pass in a single directory through dirs.

jakesmith commented on esp/services/ws_dfu/ws_dfuXRefService.cpp:
```diff
@@ -599,11 +599,25 @@ bool CWsDfuXRefEx::onDFUXRefList(IEspContext &context, IEspDFUXRefListRequest &r
 
+#ifdef _CONTAINERIZED
+        DBGLOG("CONTAINERIZED(CWsDfuXRefEx::onDFUXRefList)");
+        Owned<IPropertyTreeIterator> planesIter = getPlanesIterator("data", nullptr);
```
This k8s code may work for BM too.. would be good to unify if so.
BM has "data" planes automatically generated from the clusters, so it should in theory..

jackdelv replied:
Yes, the BM code now iterates through the 'clusters' using getPlanesIterator("data", nullptr). It partially worked where the Thor and Roxie clusters showed up properly and could be run, but additional 'clusters' (mythor_mirror, hthor_hthor, etc) were found that failed when trying to generate results.

I have removed the DBGLOG with a containerized message from the generalized code to avoid confusion.

jakesmith reviewed (commented):
Hi @jackdelv, it's looking good, I've given it a first pass, and made some comments. Please take a look.

jakesmith commented:
> I'm not sure that I correctly added the new XRef pod because everything still runs on the ECL Watch pod.

that's fine for now. We can move it to a separate esp service pod if we really want to.

jackdelv commented on dali/base/dadfs.cpp:
```diff
@@ -246,6 +246,10 @@ RemoteFilename &constructPartFilename(IGroup *grp,unsigned partno,unsigned partm
     StringBuffer fullname;
     if (findPathSepChar(name)==NULL)
         addPathSepChar(fullname.append(partdir));
+#ifdef _CONTAINERIZED
```
Removed remotePhysicalFiles and added a constructPartFilename function to CLogicalNameEntry. I had some trouble finding the dirPerPart flag in the metadata. The only examples I was able to find was using queryDirPerPart() on a StoragePlane, but when trying that it seemed the metadata was missing and it was defaulting to the return value of isContainerized().

Added lfnHash to the CLogicalNameEntry and used calcStripeNumber on the part number.

jackdelv replied:
The old constructPartFilename function had some tests for it, so I couldn't delete the function without also removing the tests. Should there now be tests for the new constructPartFilename if it does get added to CLogicalNameEntry?

jackdelv replied:
Removed the _Containerized changes and added a comment to the old constructPartFilename.

jackdelv reviewed (commented):
@jakesmith I left some responses to your comments and questions. Back to you.

jakesmith reviewed (commented):
@jackdelv - looks good overall.
Please see comments and questions.

jackdelv commented on dali/dfuXRefLib/dfuxreflib.cpp:
```diff
@@ -757,6 +760,37 @@ struct CLogicalNameEntry: public CInterface
+        }
+        expandMask(partName, pmask, partNo, max);
+
+        StringBuffer fullname;
```
Fixed.

jackdelv commented on dali/dfuXRefLib/dfuxreflib.cpp:
```diff
@@ -757,6 +760,37 @@ struct CLogicalNameEntry: public CInterface
+        Owned<IStoragePlane> plane = getDataStoragePlane(grpname, true);
+        const char * prefix = plane->queryPrefix();
+        unsigned stripeNum = calcStripeNumber(partNo+1, lfnHash, plane->numDevices());
+        bool dirPerPart = max>1?plane->queryDirPerPart():false;
```
Added file to new constructPartFilename parameters to get dirPerPart from the flags.

jackdelv commented on dali/dfuXRefLib/dfuxreflib.cpp:
```diff
@@ -942,6 +943,17 @@ static bool parseFileName(const char *name,StringBuffer &mname,unsigned &num,uns
+                        StringBuffer stripeMask;
+                        stripeMask.append(getPathSepChar(mname)).append("$P$").append(getPathSepChar(mname));
+                        // Replace part number in stripe directory with mask
+                        mname.replaceString(stripeDir, stripeMask);
```
I have restructured how it gets the dir-per-part number and added checks for the stripe number. I was able to test by changing the string before passing into the function, but I am not sure how to add to the CPPUNIT tests.

jackdelv commented on dali/dfuXRefLib/dfuxreflib.cpp:
```diff
@@ -1867,7 +1903,9 @@ class CXRefManager: public CXRefManagerBase
                         CriticalUnblock unblock(manager.logsect);
-                        unsigned short port = getDafsPort(node.endpoint(),numfails,&crit);
+                        unsigned short port = 0;
+                        if (!strieq(msg, "localhost"))
```
Added comment describing purpose of localhost check.

jackdelv commented on dali/dfuXRefLib/dfuxreflib.cpp:
```diff
@@ -2669,24 +2729,30 @@ IPropertyTree *  runXRef(unsigned nclusters,const char **clusters,IXRefProgressC
+    {
+        DBGLOG("CONTAINERIZED(runXRef)");
+        numdirs = 1;
+        Owned<IPropertyTree> storagePlane = getStoragePlane(clusters[0]);
```
I agree it makes more sense this way. I have moved the directory getting code into process.

jackdelv reviewed (commented):
@jakesmith I have responded to your comments and questions.

I think the new constructPartFilename function should be added to the CPPUNIT tests, but I am not sure how they work or the best way to extend them.

jackdelv commented on dali/dfuXRefLib/dfuxreflib.cpp:
```diff
@@ -757,6 +760,42 @@ struct CLogicalNameEntry: public CInterface
+        unsigned stripeNum = calcStripeNumber(partNo+1, lfnHash, plane->numDevices());
+
+        // Get dir-per-part # from file metadata or storage plan info
+        FileDescriptorFlags flags = static_cast<FileDescriptorFlags>(file.getPropInt("Attr/@flags"));
```
@jakesmith I added file to the parameters for constructPartFilename, but I think that may be incorrect since it uses file in the CLogicalNameEntry constructor. Which is considered best practice, initializing CLogicalNameEntry with the dir-per-part/stripe-number and adding additional members to the class or passing file into constructPartFilename and getting the information right when we need it?

jakesmith commented on dali/dfuXRefLib/dfuxreflib.cpp:
```diff
@@ -923,17 +964,60 @@ static bool parseFileName(const char *name,StringBuffer &mname,unsigned &num,uns
         char c=*name;
         if (!c)
             break;
+        if ((c=='d')&&(isdigit(name[1])))
```
This routine should really parse the filename, looking for a plane prefix match first.
Reject if no match. All files should exist in a plane.

Then check next segment for striping, and check if corresponding plane has striping, if it hasn't then complain/reject.
Then parse remainder of path, and check penultimate segment for dir-per-part pattern (i.e. only check path segment before the filename).

Because this is called a lot, it does need to be efficient.
So I would collect the data plane prefixes upfront, then initially check to see if the name has a prefix of one of them.

jakesmith commented on dali/dfuXRefLib/dfuxreflib.cpp:
```diff
@@ -923,17 +964,60 @@ static bool parseFileName(const char *name,StringBuffer &mname,unsigned &num,uns
+            ForEach(*planesIter)
+            {
+                const IPropertyTree &plane = planesIter->query();
+                if (startsWithIgnoreCase(mname.str(), plane.queryProp("@prefix")))
```
should treat paths as case sensitive

jakesmith commented on dali/dfuXRefLib/dfuxreflib.cpp:
```diff
@@ -923,17 +964,60 @@ static bool parseFileName(const char *name,StringBuffer &mname,unsigned &num,uns
             }
+
+            if (dirPerPart&&dirPerPart!=pn)
+                throw MakeStringException(-1, "dir-per-part # does not match part # of file");
```
trivial: MakeStringException is deprecated in favour of makeStringException or makeStringExceptionV

jakesmith commented on dali/dfuXRefLib/dfuxreflib.cpp:
```diff
@@ -2596,59 +2673,84 @@ class CXRefManager: public CXRefManagerBase
+                DBGLOG("CONTAINERIZED(CXRefManager::process)");
+
+                const char *storagePlaneName = clusters[i]; // clusters holds a list of storage plane names
+                Owned<IPropertyTree> storagePlane = getStoragePlane(storagePlaneName);
```
worth guarding against this being null with a suitable error.

jakesmith commented on dali/dfuXRefLib/dfuxreflib.cpp:
```diff
@@ -2596,59 +2673,84 @@ class CXRefManager: public CXRefManagerBase
+                const char *storagePlaneName = clusters[i]; // clusters holds a list of storage plane names
+                Owned<IPropertyTree> storagePlane = getStoragePlane(storagePlaneName);
+                storageDir[0] = storagePlane->queryProp("@prefix");
+                Owned<IGroup> g = queryNamedGroupStore().lookup(storagePlaneName);
```
ditto

jakesmith commented on dali/dfuXRefLib/dfuxreflib.cpp:
```diff
@@ -1517,18 +1601,14 @@ void loadFromDFS(CXRefManagerBase &manager,IGroup *grp,unsigned numdirs,const ch
                         return;
                     }
-                    constructPartFilename(grp,partno,numparts,partname,partmask,partdir,replicate,replicateoffset,rfn);
+                    lnentry->constructPartFilename(partno, replicate, rfn, file);
```
At the bottom of the loop, it cycles around unless replicate is true.
In containerized (at least for the time being) this won't make sense. The code should break out if containerized with a comment saying replication not supported in containerized.

jakesmith reviewed (commented):
@jackdelv - apologies for taking an age to get back to this.

Please see comments.
It's definitely getting closer.

jackdelv reviewed (commented):
@jakesmith I responded to all your comments and left a couple small questions. Back to you.

jakesmith reviewed (commented):
Thanks @jackdelv.
See follow up comments/replies.

Is the mirror plane being handled correctly now (being ignored) in BM?

jackdelv reviewed (commented):
@jakesmith I addressed your comments and added some tests for parseFilename to datest. I also removed calls to deprecatedConstructPartFilename in a few places to test the new function.

For the constructPartFilename tests, I am not sure how to properly judge whether or not they pass. There weren't any exceptions in the constructPartFilename calls, but also nothing validating the output (maybe it is and I just don't see it). They passed on the first run but would cause errors on a second. Is this expected?

Also, to respond to your earlier question, no, the mirror planes are not ignored at the moment. Should there be a flag added so we can ignore by type?

jackdelv commented on dali/dfuXRefLib/dfuxreflib.cpp:
```diff
@@ -757,6 +760,42 @@ struct CLogicalNameEntry: public CInterface
+        expandMask(partName, pmask, partNo, max);
+
+        // Get stripeNum from storage plane
+        Owned<IStoragePlane> plane = getDataStoragePlane(grpname, true);
```
Fixed to pass in required=false.

jakesmith commented on dali/dfuXRefLib/dfuxreflib.cpp:
```diff
@@ -917,23 +939,76 @@ static void constructPartname(const char *filename,unsigned n, StringBuffer &pn,
+    }
+
+    if (mname.isEmpty())
+        throw makeStringExceptionV(-1, "Could not find matching prefix in plane definition for file %s", filename);
```
can the caller handle exceptions? I think not, or it looks like it will cause it to bail out of too much just because of 1 malformed physical filename.

Also, this function can now either throw exception or return false - which begs question how should they be considered differently? I think throwing an exception is fine if consistent.

I would remove the boolean result, throw an exception instead of the return false, and have the caller catch, log, skip further processing of this part, and continue

jakesmith replied:
NB: The alternative would be to CLogicalNameEntry create a dirpartmask that encapsulated an abstract stripe # and dirperpart, e.g. using $S for stripe.
e.g. so it filled dirpartmask with something like "/var/lib/HPCCSystems/hpcc-data/d$S$/mydata/sub/$P$/myfile._$P$_of_$N$" , and for parseFilename to do the same..

That would be a better approach I suspect, but requires more changes to processFile.

@jackdelv - happy to discuss further.

jackdelv replied:
I removed the boolean returns, and made exception throwing more consistent. The caller will now catch any error from parseFileName and log it with the manager.

I think it would fairly easy to build up a dirpartmask and return that alongside mname. Should that be done here, or in a follow up PR?

jakesmith replied:
let's leave for a subsequent PR. Can you open a separate JIRA to follow?

jakesmith commented on dali/dfuXRefLib/dfuxreflib.cpp:
```diff
@@ -917,23 +939,74 @@ static void constructPartname(const char *filename,unsigned n, StringBuffer &pn,
+                name++;
+                while (*name&&isdigit(*name))
+                    name++;
+                mname.append("d$P$");
```
$P$ should be left as meaning the part # only.

Horrible, but if you look at how the result parseFilename 'mname' is used by the caller (i.e. 'orphanname'), it is expecting it to map it against logical file entries (lines 1702-1705), based on CLogicalNameEntry objects created by parseFile (line 1485) whose key value (in manager.logicaldirmap) is 'dirpartmask'. And, 'dirpartmask' is based on the logical files @directory and it's @partmask only.

IOW, it has no concept of stripe or partPerDir at this stage.
It is trying to track physical files found that, in containerized, likely will have stripe and dir per part onto logical file entries that (where stripe nor dirperpart are part specific and the logical file is abstract from that).

So, parseFilename should not be trying to insert stripe or dirperpart into into the returned mname here. It should parse it only.
It may be useful to capture the info. and return the strip # and dir-per-part info as separate info (for future use *maybe*), but adding to the mname will cause every found part here, to fail to match up with the logical file it corresponds to.

jakesmith commented:
> For the constructPartFilename tests, I am not sure how to properly judge whether or not they pass. There weren't any exceptions in the constructPartFilename calls, but also nothing validating the output (maybe it is and I just don't see it). They passed on the first run but would cause errors on a second. Is this expected?
> 

Unit tests should validate the output - using CPPUNIT_ASSERT_MESSAGE.

> Also, to respond to your earlier question, no, the mirror planes are not ignored at the moment. Should there be a flag added so we can ignore by type?

We should not be performing XREF on the mirror planes (only relevant for BM), at least for now (as per how it used to be behave),
and since it is now standarizing on planes, it must be careful it doesn't XREF them unintentionally.
I previously made a comment that it could add a flag when mirror planes are created, so that XREF could 'see' them and ignore them. I think that still should be done.

jakesmith reviewed (commented):
Hi @jackdelv , please see comments. Happy to discuss any if that would help.

jackdelv commented on dali/dfuXRefLib/dfuxreflib.cpp:
```diff
@@ -917,23 +939,76 @@ static void constructPartname(const char *filename,unsigned n, StringBuffer &pn,
+            while (*tailSlash&&*tailSlash!='/')
+                tailSlash--;
+            const char *d = tailSlash-1;
+            for (int i=1;*(d)&&isdigit(*d);i*=10,d--)
```
Fixed

jackdelv commented on dali/dfuXRefLib/dfuxreflib.cpp:
```diff
@@ -917,23 +939,76 @@ static void constructPartname(const char *filename,unsigned n, StringBuffer &pn,
+
+            // Check for dir-per-part number
+            const char *tailSlash = cur;
+            while (*tailSlash&&*tailSlash!='/')
```
Fixed

jackdelv commented on dali/dfuXRefLib/dfuxreflib.cpp:
```diff
@@ -917,23 +939,76 @@ static void constructPartname(const char *filename,unsigned n, StringBuffer &pn,
+                if (dirPerPart!=pn)
+                    throw makeStringExceptionV(-1, "Dir-per-part # does not match part # of file %s", filename);
+
+                mname.append((d+1)-name, name).append("$P$").append(cur-tailSlash, tailSlash);
```
No longer adding masks to mname.

jackdelv commented on dali/dfuXRefLib/dfuxreflib.cpp:
```diff
@@ -917,23 +939,76 @@ static void constructPartname(const char *filename,unsigned n, StringBuffer &pn,
+                    throw makeStringExceptionV(-1, "In storage plane definition numDevices>1, but no stripe sub-directory found in file %s", filename);
+                name++;
+                while (*name&&isdigit(*name))
+                    name++;
```
Added more validations and errors for each case.

jackdelv commented on dali/datest/datest.cpp:
```diff
@@ -3317,17 +3320,107 @@ void testlockprop(const char *lfn)
     printf("done\n");
 }
 
+void TestParseFileName()
```
Moved tests to dfuxreflib.cpp

jackdelv commented on dali/dfuXRefLib/dfuxreflib.cpp:
```diff
@@ -917,23 +939,76 @@ static void constructPartname(const char *filename,unsigned n, StringBuffer &pn,
-    replicate = setReplicateDir(name,nonrepdir,false);
-    if (replicate) 
-        name = nonrepdir.str();
+    if (!isContainerized())
```
Fixed.

jackdelv commented:
@jakesmith I added a flag "@copy" to the mirror storage planes. They are now ignored if copy is true.

jakesmith commented on dali/dfuXRefLib/dfuxreflib.cpp:
```diff
@@ -3010,3 +3032,158 @@ IPropertyTree * RunProcess(XRefCmd cmd, unsigned nclusters,const char **clusters
+
+        // Standard file name with multiple parts
+        parseFileName("/var/lib/HPCCSystems/hpcc-data/test/myname._1_of_3", mname.clear(), num, max, stripeNum, dirPerPart, replicate);
+        CPPUNIT_ASSERT_EQUAL_STR("/var/lib/HPCCSystems/hpcc-data/test/myname._$P$_of_3",mname.str());
```
trivial: (no need to change) - maybe slight overkill for new macro that converts each arg. to std::string's I think.
Could could be:
```
CPPUNIT_ASSERT(strsame("/var/lib/HPCCSystems/hpcc-data/test/myname._$P$_of_3", mname));
```

jakesmith commented on dali/dfuXRefLib/dfuxreflib.cpp:
```diff
@@ -3010,3 +3032,158 @@ IPropertyTree * RunProcess(XRefCmd cmd, unsigned nclusters,const char **clusters
+        CPPUNIT_ASSERT_EQUAL_MESSAGE("Replicate is incorrect",true,replicate);
+
+        // A file without a storage plane throws an exception
+        try {
```
trivial/style: BSD style where open/close blocks are on a newline is the preferred style (+ new code).

jakesmith commented on dali/dfuXRefLib/dfuxreflib.cpp:
```diff
@@ -958,16 +964,26 @@ static bool parseFileName(const char *name,StringBuffer &mname,unsigned &num,uns
             if (plane.getPropInt("@numDevices") > 1)
             {
-                if (*name&&*name!='d')
+                name++;
```
this is skipping the PATHSEPCHAR that the code established was present on line 964.
Logically this name + 1 should be before the "if (plane.getPropInt("@numDevices") > 1)" block, which is how it was afaics, prior to most recent changes.
i.e. I think line 972 should be deleted and line 968 should be: 
```
name += prefixLen + 1;
```

jakesmith commented on dali/dfuXRefLib/dfuxreflib.cpp:
```diff
@@ -994,17 +1010,17 @@ static bool parseFileName(const char *name,StringBuffer &mname,unsigned &num,uns
+            for (int i=1;d!=name&&isdigit(*d);i*=10,d--)
                 dirPerPart += (*d-'0')*i;
+            // If a dir-per-part number was found, check that it matches the part number
             if (*d&&*d=='/')
```
similar to a prior comment - I don't think it is valid here to be testing *d for '\0' - that should never be possible. If it null, it suggests it has walked past the beginning of the input string and into memory where it has hit a 0.

jakesmith reviewed (commented):
@jackdelv - looks very close, please see new minor comments.

jackdelv reviewed (commented):
@jakesmith I fixed the minor comments you left and opened a [JIRA](https://hpccsystems.atlassian.net/browse/HPCC-33138) to follow up on one of the comments. Should I open any additional JIRAs to follow on with this?

jakesmith commented on dali/dfuXRefLib/dfuxreflib.cpp:
```diff
@@ -914,51 +941,120 @@ static void constructPartname(const char *filename,unsigned n, StringBuffer &pn,
+    else
+        replicate = false; // Replicate dir is not supported in containerized environment
+
+    Owned<IPropertyTreeIterator> planesIter = getPlanesIterator("data", nullptr);
```
This code should be revisited after this PR is merged. I've opened and linked : https://hpccsystems.atlassian.net/browse/HPCC-33151

jakesmith reviewed (commented):
@jackdelv - looks good. Please squash (but don't rebase) for a final look. Can also convert from Draft PR

Have you done some testing of XREF with this PR in bare-metal, to ensure it hasn't introduced any issues?

jackdelv commented:
@jakesmith I did some testing with logical files to compare the results of running XRef and made sure nothing changed in bare-metal. Is there a better way than just comparing runs of the two versions?

I did add an additional flag to remove the plane hthor_myeclagent from showing up in the list of XRef options because it had the message "Not Found" when trying to generate the results. I also added SuperFiles back to the list of options for bare-metal because it was being used to turn on SuperFile checking at saxref.cpp:2084.

jakesmith commented on dali/base/dafdesc.cpp:
```diff
@@ -3576,10 +3576,19 @@ void GroupInformation::createStoragePlane(IPropertyTree * storage, unsigned copy
     IPropertyTree * plane = storage->addPropTree("planes");
+
+    // Revisit: Ignore hthor planes when running XRef on storage planes
+    if (startsWith(name, "hthor"))
```
there *could* be a plane name starting with hthor that isn't an hthor group, so this is a bit of kludge.
It would be cleaner to pass in the GroupType to GroupInformation::createStoragePlane and test against 'grp_hthor'

jackdelv replied:
Changed to use GroupInformation::groupType member to add flag.

jakesmith reviewed (commented):
@jackdelv - looks good afaics, please squash and rebase to resolve clash

jakesmith commented:
> @jakesmith I did some testing with logical files to compare the results of running XRef and made sure nothing changed in bare-metal. Is there a better way than just comparing runs of the two versions?
> 

have you done some sanity checks to deliberately test it produces xref results for orphan parts and/or detached files in both BM and cloud?
..by e.g.:

1) manually deleting some physical parts from existing logical files, 
2) deleting the meta data for a logical file - leaving the physical files to be unassociated.
- [1] daliadmin translatetoxpath <logicalfilename>
- [2] daliadmin . delete <result from [1]>

I can't recall where, but somewhere I think there's a check for a time difference, where it will ignore orphans if they're new'ish.
So you might either need to find that and config/change that delay, or make the orphan files old

jackdelv commented:
@jakesmith I did do some sanity testing to make sure it was able to find found, orphan, and lost files. In bare-metal, the results looked the same and seemed to be correct. 

The only thing I noticed was in the containerized version, the status message after running XRef was "generated" as opposed to in bare-metal it was "generated[3]" where I think 3 refers to the number of logical files that were found.

Also, now any storage planes defined will show up as an option to run in XRef on bare-metal, but when ran it returned a status message of "Not Found".

jakesmith commented:
>The only thing I noticed was in the containerized version, the status message after running XRef was "generated" as opposed to in bare-metal it was "generated[3]" where I think 3 refers to the number of logical files that were found.

I think worth hunting down the reason for the difference, in case more significant that it 1st appears.

>Also, now any storage planes defined will show up as an option to run in XRef on bare-metal, but when ran it returned a status message of "Not Found".

Do you mean manually added external storage planes added to BM (i.e. the new ability to do so in HPCC-32724) ?
If so, happy to ignore that one for now, but should open a new JIRA to revisit the cause.

jakesmith reviewed (commented):
@jackdelv - please see reply. I thin worth hunting the down the cause of the difference re. the "generated" different before we commit this PR.

jakesmith commented:
> > The only thing I noticed was in the containerized version, the status message after running XRef was "generated" as opposed to in bare-metal it was "generated[3]" where I think 3 refers to the number of logical files that were found.
> 
> I think worth hunting down the reason for the difference, in case more significant that it 1st appears.

@jackdelv - just adding notes from our offline discussion. The cause there are currently two distinct xref starting points. The standard one is via the sasha xref service, the other is via the esp service (I suspect the esp route should be removed, but not that's s separate issue).
The difference at the moment is the configuration. In bare-metal environment it is defaulting to useSasha, via 'xrefEclWatchProvider="true"' in sasha in the environment.xml. In containerized it is not (but should be).

This points to an issue with the sasha xref's config in cloud. These options should be accessible if set under 'sasha:' in the values yaml.
At the moment that won't work because the server service code is geared only toward BM setup.
It needs code similar to other sasha services to differentiate where it gets its config from, e.g.:
```
#ifdef _CONTAINERIZED
    Linked<IPropertyTree> config = serverConfig;
#else
    Owned<IPropertyTree> config = serverConfig->getPropTree("DfuXRef");
    if (!config)
        config.setown(createPTree("DfuExpiry"));
#endif
```
Currently, there is also code that looks at 'clusters' by connecting to Dali and looking at /Environment/Software.
It should be possible now in both BM and containerized to get this info from planes without connecting to Dali.

jakesmith commented on dali/sasha/saxref.cpp:
```diff
@@ -803,7 +803,12 @@ class CNewXRefManager: public CNewXRefManagerBase
+            rootdir.set(getStoragePlane(_clustname)->queryProp("@prefix"));
+            assert(!rootdir.isEmpty());
+        }
+        else if (basedir.length()==0) {
```
trivial/style: mixture of K&R vs Allman style. Best to keep it consistent with existing function style or change them all within a function, and for new code Allman style.

jakesmith commented on dali/sasha/saxref.cpp:
```diff
@@ -2086,14 +2091,9 @@ class CSashaXRefServer: public ISashaServer, public Thread
         StringArray groups;
         StringArray cnames;
-        clustersToGroups(conn->queryRoot(),list,cnames,groups,NULL);
+        planesToGroups(list,cnames,groups);
```
I am not sure about this change.
In bare-metal, we have cluster names and their groups may be different, clustersToGroups is mapping the cluster list provided to runXRef to their underlying groups.

In containerized, we are seeking to run xref per plane. There shouldn't be a selection based on 'cluster' which are [queue] targets in containerized, because that is somewhat meaningless in terms of the disconnect storage they use.
There will always be a 1:1 mapping of plane name -> group name (as created automatically by dafdesc.cpp: initializeStoragePlanes).

So I think runXRef in bare-metal can/should continue to provide from a cluster perspective, and therefore needs to map cluster->group etc. (NB: the group overlap stuff between lines 758-766 doesn't make sense in containerized, we should make it conditional, add comments ..).
It is probably easiest to leave the clustersToGroups call as it was for BM for now, and in containerized assert that the 'clustcsl' passed into runXRef contains plane names.

In the longer run, both BM and containerized should be 'plane' centric.. you should select a storage plane, not a cluster in both cases, and this code should be cleared up to remove it's references and tracking of 'clusters'.

jackdelv replied:
Changed to revert bare-metal code. Changed planesToGroups to parse clustcsl and assert it matches a storage plane.

jakesmith commented on helm/hpcc/values.yaml:
```diff
@@ -511,7 +511,9 @@ sasha:
 
-  xref: {} # NB: no properties defined, use defaults
+  xref:
+    eclwatchProvider: true
```
I suspect it might be better to default this to true in the code in k8s, and remove from here.
I suspect it will soon be the only option.

jakesmith reviewed (commented):
@jackdelv - please see comments.

jackdelv commented:
@jakesmith I responded to your comments. I added parseFileName to cFileDesc::addFile. In order to get the logical file name for cFileDesc, I changed parseFileName to not include the prefix in the returned file name. 

There were a few places that I ran into issues where I had to add a check for isContainerized. I made some comments in the latest commit about some places in the code (dautils.cpp:3810,saxref.cpp:1215).

jakesmith commented on dali/sasha/saxref.cpp:
```diff
@@ -803,7 +814,11 @@ class CNewXRefManager: public CNewXRefManagerBase
         rawgrp->queryNode(0).endpoint().getHostText(gname.clear());
```
I've no idea what this is for/doing..
I suspect we should at least put it in an if (!isContainerized())..
But can you check how it behaves in containerized setup and then discuss?

Also the cover above this.. it looks like it's scanning for other groups that contain the target group... 
Probably trying to cope with situations (on BM) where there are overlapping clusters.
If so need to be careful not to cause strange behaviour in a containerized setup, when most groups will be full of localhosts.. can you check what effect it has in a containerized setup with a couple of data planes?

jakesmith reviewed (commented):
@jackdelv - please see comments.

I think close now, a few things to check - and some 'revisit's to come back to, but hopefully we can push this over the line in the next couple of days.

jakesmith commented on dali/sasha/saxref.cpp:
```diff
@@ -2178,14 +2194,19 @@ class CSashaXRefServer: public ISashaServer, public Thread
         const char *clusters = props->queryProp("@clusterlist");
         StringBuffer clusttmp;
-        if (props->getPropBool("@checkSuperFiles",true)) {
+        if (props->getPropBool("@checkSuperFiles",isContainerized()?false:true)) {
```
Superfiles won't have their own plane, they could container files that live on multiple planes - I think XREF's "checksuperfiles" should be checking the integrity of super files - so very different from checking meta against physical.

It should work in containerized too, but it shouldn't be looking it up as a plane - nor should BM be looking it up a cluster/group. Where is it looking it up as a plane a failing?

jakesmith reviewed (commented):
@jackdelv - looks v. close. There's a merge clash - can you squash and rebase to candidate-9.10.x

jakesmith reviewed (commented):
@jackdelv - Please see comments.

I ran a test system, created a single file, and deleted the meta data, then ran xref.
Noticing a couple of issues so far:

1) scanDirectories, ~line 1002, it's trying to scan hpcc-mirror (the replicate dir). It shouldn't be. It should finish after the 1st scanDirectory.

2) Stepping through the code, looked like it was producing an orphaned report, but eclwatch - but Eclwatch isn't displaying the reports afaics ? Eclwatch reports : 
![image](https://github.com/user-attachments/assets/36dddaab-0740-4809-9014-f458c151c175)

jakesmith commented on dali/sasha/saxref.cpp:
```diff
@@ -2086,14 +2106,28 @@ class CSashaXRefServer: public ISashaServer, public Thread
+                return;
+            }
+            clustersToGroups(conn->queryRoot(),list,cnames,groups,NULL);
+        }
```
Related.. lines that say serverConfig->....("DfuXRef/@<someprop>" ... need changing.
Containerized config will not have a "DfuXRef" node. It should pick these properties up at the top level.

The run() command defines 'props' correctly, based on whether containerized or BM.
I suggest persisting 'props' as a member, and using it instead of serverConfig (and without the "DfuXRef/") then that code will work for both containerized and BM.

jackdelv committed:
Fix files with dirPerPart not being found
- expandMask wasn't taking stripeNum and dirPerPart into account when creating the physical file name from the part mask
- scanDirectory was calling itselg recursively on the dirPerPart directories causing files parts to be under different cDirDesc which prevented them from being logged as found.
